### ISSUE TITLE
synchronization bug (create -> start -> kill) 

### ISSUE THREAD

**Issue by chavafg** [clearcontainers/runtime#430]
This is a duplicate of: https://github.com/01org/cc-oci-runtime/issues/828, I am opening this issue to have better tracking in 3.0. 

Tried to execute https://github.com/01org/cc-oci-runtime/blob/master/tests/integration/docker/kill.bats 
with CC3.0. Second test still fails just as in 2.x. 

Seems like it is a race condition since @grahamwhaley ran some tests in baremetal and couldn't reproduce it. I ran my tests under a VM with 2 CPUs and 4 GB of ram.

**Comment by sboeuf** [comment]
I have started to debug this. I can reproduce the issue described here:
```bash
for i in `seq 1 15`; do docker run -dti --name bla$i ubuntu bash -c "trap \"exit 15\" SIGUSR1 ; while : ; do sleep 1; done" && docker kill -s SIGUSR1 bla$i ; done
```
If we check the state of the containers with
```
docker ps -a
```
Some of the containers have exited and some others are still running.

Interesting thing is that I have added an extra `sleep 1` after we run the container workload, and all the kill signals seem to reach the container properly because all the containers exit with the code 15.
Here is the other script:
```
for i in `seq 1 15`; do docker run -dti --name bla$i ubuntu bash -c "trap \"exit 15\" SIGUSR1 ; while : ; do sleep 1; done" && sleep 1 && docker kill -s SIGUSR1 bla$i ; done
```

So basically my understanding is that the container needs some time to be able to handle the signal. Need to investigate shim/proxy to figure out what is going wrong here.

**Comment by devimc** [comment]
the problem is because of the workload has not been executed (has not cpu time) when agent returns reporting success, I found a fix for 2.x (and I think could be the same for 3.0) but it implies to change cpu affinity and scheduling policy in the VM in order to guarantee the workload has been executed, or we can just put a sleep before returning from ```STARTCONTAINER``` in the agent

**Comment by grahamwhaley** [comment]
Hi @devimc So we can discuss further, **exactly** what is the root of the problem - for instance, I think it is that the VM/agent is up and running, but the workload itself is not - so when the runtime posts a signal to the agent to send to the workload, we get 'stuck' in the VM/agent - maybe?
Maybe a sequence (diagram or just text listing) of a working and non-working flow could help us understand here. That might help answer questions like:
Q. Do we think this is something to fix in the agent - that it cannot deliver a signal until it knows the workload has started running?

As you know, changing the cpu affinity and scheduler policy in the VM kernel/mini-OS is not I think the correct solution (but did/does give us a big clue to the problem :-)

**Comment by jodh-intel** [comment]
It would probably be useful to collect and attach some logs from runtime+proxy after configuring the agent for debug (https://github.com/clearcontainers/runtime/blob/master/docs/debug-agent.md).

**Comment by devimc** [comment]
@grahamwhaley I agree, changing cpu affinity and/or scheduler is not the solution.
the problem is that the workload has not been executed (consume cpu / uploaded to cpu) when the kill signal arrives (I guess the process was just scheduled) because the execution of a process depends of the scheduler, the only way I found to address this issues is adding a ```sleep``` before return success in the ```newcontainer ``` command (agent) , but this is not the best solution.
I guess the ```sleep``` allows to the workload be executed (consume cpu / uploaded to cpu)

**Comment by devimc** [comment]
no idea how to fix this issue, the more near approach I tried was https://github.com/devimc/agent/tree/issue430,  but the issue still occurs with less frequency

**Comment by devimc** [comment]
I can't reproduce this issue using busybox image  😕 

```
for i in `seq 1 15`; do docker run -dti --name bla$i busybox sh -c "trap \"exit 15\" SIGUSR1 ; while : ; do sleep 1; done" && docker kill -s SIGUSR1 bla$i ; done
```

**Comment by devimc** [comment]
I tried to reproduce this issue in a VM with the linux-containe kernel using ```runc``` as the default runtime and the issue does not occurs, with this experiment I can discard that this issues is related to linux-container kernel

```
/usr/bin/qemu-system-x86_64 -machine pc,accel=kvm,kernel_irqchip,nvdimm -cpu host -device \
nvdimm,id=nv0,memdev=mem0 -object \
memory-backend-file,id=mem0,mem-path=container.img,size=524288000 \
-m 2048M,slots=2,maxmem=16998M -smp 4,cores=4,threads=1,sockets=1 \
-vga none -nographic -no-user-config -nodefaults \
-kernel vmlinux-4.9.35-66.1.container -append root=/dev/pmem0p1 \
rootflags=dax,data=ordered,errors=remount-ro rw rootfstype=ext4 \
tsc=reliable no_timer_check rcupdate.rcu_expedited=1 i8042.direct=1 i8042.dumbkbd=1 \
i8042.nopnp=1 i8042.noaux=1 noreplace-smp reboot=k panic=1 \
console=hvc0 console=hvc1 initcall_debug iommu=off cryptomgr.notests net.ifnames=0 console=ttyS0 \
-device virtio-serial-pci,id=virtio-serial0 -chardev stdio,id=charconsole0,signal=off -device \
virtconsole,chardev=charconsole0,id=console0 -net nic,model=virtio -net \
user,id=net0,net=192.168.76.0/24
```
```
# qemu-system-x86_64 --version
QEMU emulator version 2.6.2 (qemu-2.6.2-8.fc24), Copyright (c) 2003-2008 Fabrice Bellard
```

**Comment by jodh-intel** [comment]
A few more data points...

- I cannot recreate the problem with the `alpine` image.
  Alpine uses busybox for `/bin/sh` but ubuntu has `/bin/sh -> /bin/dash`.

- I created an ubuntu image with busybox installed and was unable to recreate the issue
  (although it can occasionally take a second or so for `docker ps -a` to show all containers have stopped).

- I also tested the ubuntu image using `bash` and, oddly, the behaviour is better than `dash`, but with `bash` a few containers get left around after the test (1-3 in my testing). A subsequent `docker kill` for the stragglers does kill them.

**Comment by devimc** [comment]
issue https://github.com/clearcontainers/agent/issues/105  could be related with this one

**Comment by devimc** [comment]
after analyze runc and cc-agent behaviour I think this issue could be related to ```create -> start``` process, cc-agent spawns the workload at ```start``` stage, while runc at ```create``` stage giving to the workload more time to be executed

**Comment by devimc** [comment]
I did a local test moving ```start``` process to ```create``` (same as ```run```) and I can't reproduce the issue (fixed!!), but I guess there is a reason of why we don't execute the workload at ```create```. does this imply a re-design in CC 3.0?

```
diff --git a/create.go b/create.go
index 75f7887..c46240c 100644
--- a/create.go
+++ b/create.go
@@ -218,6 +218,11 @@ func createContainer(ociSpec oci.CompatOCISpec, containerID, bundlePath,
                return vc.Process{}, err
        }
 
+       c, err = vci.StartContainer(podID, containerID)
+       if err != nil {
+               return vc.Process{}, err
+       }
+
        return c.Process(), nil
 }
 
diff --git a/start.go b/start.go
index a2ca286..ba2ce43 100644
--- a/start.go
+++ b/start.go
@@ -66,10 +66,5 @@ func start(containerID string) (vc.VCPod, error) {
                return vci.StartPod(podID)
        }
 
-       c, err := vci.StartContainer(podID, containerID)
-       if err != nil {
-               return nil, err
-       }
-
-       return c.Pod(), nil
+       return nil, nil
 }
```

**Comment by devimc** [comment]
executing the workload at the ```create``` stage we could fix also https://github.com/containers/virtcontainers/issues/356

**Comment by jcvenegas** [comment]
@devimc so we need a way to ensure startPod guarantee to the continer process is really running ? or implement create in a similar way runc does by running the contianer process in start and pause it ?

**Comment by devimc** [comment]
@jcvenegas  same as runc, spawning the workload at ```create``` stage

**Comment by jodh-intel** [comment]
The OCI spec is clear on this point: the workload must be run at `start` time:

- From https://github.com/opencontainers/runtime-spec/blob/master/runtime.md#create

  "process.args MUST NOT be applied until triggered by the start operation. "

- From https://github.com/opencontainers/runtime-spec/blob/master/runtime.md#start

  "This operation MUST run the user-specified program as specified by process."

afaik `runc` interprets this as putting the workload in a "ready-to-run" state in `create`. Is that your understanding @devimc / @sameo?

**Comment by grahamwhaley** [comment]
I will note - even if we do find that moving the start point around between `create` and `start` 'fixes' this - I still don't think I've seen an explanation of what has actually got stuck where, and why - such as, exactly which function/syscall has gotten stuck or failed in which program (runtime, agent etc.), and why?  Until we know that, I'd be very hesitant to call it 'fixed' - it would be more 'hidden', and may come back to bite us. I cannot see clearly in my mind how not having the workload process run yet makes something (like delivering a signal say) fail...

**Comment by jodh-intel** [comment]
Agreed. Can we get some logs for a failing run for all components (runtime+vc, proxy, shim, agent)?

**Comment by devimc** [comment]
@jodh-intel @grahamwhaley  @sameo 

```
$ for i in `seq 1 5`; do docker run -dti --name bla$i ubuntu bash -c "trap \"exit 15\" SIGUSR1 ; while : ; do sleep 1; done" && docker kill -s SIGUSR1 bla$i ; done
bbe78ac78f2f5d38043bc3c47c7ebc072735a80dd9e10356dba186bb83ce5415
bla1
be39f3e99e362530fac4b5b988c41556c7937e48c35dff744af7c74d5c33c224
bla2
ce46b8f4f0712768b8203ecf78c551473a011fa7354e764525c6184dd0f8803f
bla3
16cfd40b52c57b3383bc3d478a0b2a03174824f85bd3457b51de84f9b9ccbb39
bla4
41095d8d5edc9c0db4c77b2f933cd65578d6b5c08371f4096c8fc6548d615705
bla5
```

some containers are still up and running 
```
$ docker ps -a
CONTAINER ID        IMAGE               COMMAND                   CREATED             STATUS                       PORTS               NAMES
41095d8d5edc        ubuntu              "bash -c 'trap \"ex..."   28 seconds ago      Exited (15) 5 seconds ago                        bla5
16cfd40b52c5        ubuntu              "bash -c 'trap \"ex..."   36 seconds ago      Exited (15) 12 seconds ago                       bla4
ce46b8f4f071        ubuntu              "bash -c 'trap \"ex..."   46 seconds ago      Up 36 seconds                                    bla3
be39f3e99e36        ubuntu              "bash -c 'trap \"ex..."   52 seconds ago      Exited (15) 31 seconds ago                       bla2
bbe78ac78f2f        ubuntu              "bash -c 'trap \"ex..."   59 seconds ago      Up 52 seconds                                    bla1
```

[cc-proxy.txt](https://github.com/clearcontainers/runtime/files/1296523/cc-proxy.txt)
[runtime.log.xz.txt](https://github.com/clearcontainers/runtime/files/1296531/runtime.log.xz.txt) (rename  to runtime.log.xz)

**Comment by devimc** [comment]
I found the critical section where the "init" process exec()s the workload, I tried to improve it but the issue still occurs

```
diff --git a/vendor/github.com/opencontainers/runc/libcontainer/container_linux.go b/vendor/github.com/opencontainers/runc/libcontainer/container_linux.go
index 186faab..efe0c88 100644
--- a/vendor/github.com/opencontainers/runc/libcontainer/container_linux.go
+++ b/vendor/github.com/opencontainers/runc/libcontainer/container_linux.go
@@ -255,10 +255,11 @@ func (c *linuxContainer) exec() error {
                return err
        }
        if len(data) > 0 {
-               os.Remove(path)
-               return nil
+               return fmt.Errorf("cannot start an already running container")
        }
-       return fmt.Errorf("cannot start an already running container")
+
+       os.Remove(path)
+       return nil
 }
 
 func (c *linuxContainer) start(process *Process, isInit bool) error {
diff --git a/vendor/github.com/opencontainers/runc/libcontainer/standard_init_linux.go b/vendor/github.com/opencontainers/runc/libcontainer/standard_init_linux.go
index 580b3fe..a42016b 100644
--- a/vendor/github.com/opencontainers/runc/libcontainer/standard_init_linux.go
+++ b/vendor/github.com/opencontainers/runc/libcontainer/standard_init_linux.go
@@ -170,9 +170,7 @@ func (l *linuxStandardInit) Init() error {
        if err != nil {
                return newSystemErrorWithCause(err, "openat exec fifo")
        }
-       if _, err := unix.Write(fd, []byte("0")); err != nil {
-               return newSystemErrorWithCause(err, "write 0 exec fifo")
-       }
+       defer unix.Write(fd, []byte("E"))
        if l.config.Config.Seccomp != nil && l.config.NoNewPrivileges {
                if err := seccomp.InitSeccomp(l.config.Config.Seccomp); err != nil {
                        return newSystemErrorWithCause(err, "init seccomp")
```

I'll try to "start" the container at `create` time and "exec" it at `start` time same as runc

**Comment by grahamwhaley** [comment]
Hi. OK, I think I've worked out why some containers get the SIGUSR1 and exit, and some don't.
I believe we have a race in the test, and not in the runtime.
The race is between the `docker kill -s SIGUSR1` and the `trap "exit 15" SIGUSR1`.
The race is that the `docker kill` can post the SIGUSR1 to the running `bash`, before `bash` has executed the `trap`. As `bash` is the 'container-init pid1' for the namespace it is in, it carries the `SIGNAL_UNKILLABLE` flag in the kernel. That is, it will ignore any signal, even ones that terminate, unless it has registered an interest in them. This is handled in the kernel. The signal will not even be posted to the process.

The fix would be to have the test script in the container synchronise with the test program - that is, defer the `docker kill` until we know the `trap` has run. One way would be to have the bash script to drop a known string in a known file inside the container when it is ready, and have the test script look for and verify the existence of that file in the container (possibly using `docker exec cat` for instance) before issuing the kill.

The proof: is slightly tricky, as it relies on instrumenting the kernel to see what is going on. I did that, and the logs are below. They are not quite perfect - I could add a log of the SIGUSR1 handler being registered - but, I think we have proof enough. One way to show up the effect of the SIGNAL_UNKILLABLE wrt SIGUSR1/Terminating signals is to replace the SIGUSR1 in the `kill` line with a `SIGUSR2`. `bash` will not register an interest in SIGUSR2, but as `bash` is in UNKILLABLE init state, the SIGUSR2 will never be delivered, and none of the containers run will exit - they all keep running.

Logs:
Here is a snippet of (annotated) kernel log from a working case:
```
do_execve...(161:/bin/bash)
 is UNKILLABLE
clone()
do_fork()
copy_process()
 cp(current is UNKILLABLE)
 do_fork() generated 4
do_execve...(164:/bin/sleep)
 is not UNKILLABLE
Sending sig10 to pid 161(161)(bash)
 pid 161(161)(bash) in state [1]
woke up pid 161(161)(bash) to get sig10
 pid 161(161)(bash) in state [0]
sig10 in get_signal()
sig10 is not DFL, run handler
sig10 get_signal is done
: do_signal sig10 fire handle_signal
:  handle_signal(sig10)
 sig10 delivered to pid 161(1)(bash)
:  handle signal finished!
```
Here you can see the `exec` of the `/bin/sleep` before the sig10 (SIGUSR) delivery - thus, showing that we have passed the `trap` in the bash script (it is before the sleep). And then we see the 'not DFL' (non default) signal handler get invoked, which triggers the `exit 15`.

Here is the failing case:
```
do_execve...(161:/bin/bash)
 is UNKILLABLE
Sending sig10 to pid 161(161)(bash)
 sigmask 0
 pid 161(161)(bash) in state [0]
woke up pid 161(161)(bash) to get sig10
 sigmask 0
 pid 161(161)(bash) in state [0]
sig10 in get_signal()
sig10 not killable, skip
clone()
do_fork()
copy_process()
 cp(current is UNKILLABLE)
 do_fork() generated 5
do_execve...(165:/bin/sleep)
```

Here we see the sig10 being delivered before the `exec` of the `/bin/sleep` - and the inference here (which I could go prove probably if we really need) is that the `trap SIGUSR1` has not been executed yet either. Hence, we do not find a non-default sig handler to invoke, and as the pid1 `bash` is a container-init and has SIGNAL_UNKILLABLE, we see the sig10 get dropped in the kernel (not delivered to bash).

Summary - there is no runtime race here - the runtime successfully launched the workload (`bash`) up to the fully running point before the signal was attempted to be delivered. The bug is that the workload itself takes a finite time to get ready to handle that signal, and as we are firing the signal at the init process of the workload, if it has not registered an interest in the signal yet, it will not be delivered by the kernel.

/cc @jodh-intel @devimc - if you both agree, then I think we can close this and the Issue over on CC2.x - and/or, fix the test if there is one to fix?

**Comment by grahamwhaley** [comment]
A side note - if you want to re-create the issue using `runc` for instance, put a `sleep 10;` before the `trap` in the workload command - then `runc` will also try to deliver the SIGUSR1 before bash has registered an interest in it, and you should find all the containers still running. It is just a 'timing problem' with the test, that `cc-runtime` probably shows up as it is 'heavier'.

**Comment by devimc** [comment]
@grahamwhaley  let's close it :)
thanks

**Comment by jodh-intel** [comment]
I agree too - good debug btw!

**Comment by grahamwhaley** [comment]
OK, thanks.
Just before we close this then, to check with QA, @chavafg @GabyCT - do we have any existing tests that do this sig/exit return test that we need to go fix up. If so, let's fix them before closing this issue (or open another issue for the fix, and close this might be better). If not, then we can just close this.

**Comment by devimc** [comment]
we already have a test with a sleep https://github.com/clearcontainers/tests/blob/master/integration/docker/kill_test.go#L87  😄

**Comment by grahamwhaley** [comment]
@devimc - but I think that sleep is not the one we need.
That sleep is to wait for the whole signal thing to be processed and the container to be wrapped up.
What we need is another sleep here:
https://github.com/clearcontainers/tests/blob/master/integration/docker/kill_test.go#L78
between the 'run' and the 'kill' commands to allow the container to process the 'trap' call and be ready to handle the signal.
@sameo and I will have a look at that now (as that test is failing for @sameo on CI)

**Comment by amshinde** [comment]
@sboeuf Fix has been pushed for this : https://github.com/clearcontainers/tests/pull/535

**Comment by sboeuf** [comment]
@grahamwhaley @amshinde my bad, I have misread/misunderstood the root cause. So you're basically saying that we have this behavior because we have the `-d` option, right ? Because all the commands are run as detached, we go to the next one without waiting for the first one to be complete (basically making sure that `trap` has registered the signal properly).
One question @grahamwhaley, you mention that we can make `runc` failing if we add `sleep 10;` before the trap call, but I assume it should work with sleep 1, right ?

**Comment by grahamwhaley** [comment]
Hi @sboeuf.
 We had the behavior because the test did not wait or check that the workload had executed the `trap` before it tried to send it SIGUSR1. If the SIGUSR1 is sent before the workload has registered the `trap` then the SIGUSR1 is ignored as the `bash` workload is classified as the 'container-init' (pid1 of the container pid namespace), so it does not get sent termination signals unless it has registered to do so.

So, not a bug in the runtime, nothing to do with `-d`. The runtime had successfully launched the workload, and then successfully sent the SIGUSR1 - it's just that the workload had not executed far enough to run its `trap` by that point, and that is out of the control of the runtime.

As for `runc`, well, the timing is just different - different enough that we never seem to run into the small race window (between SIGUSR1 delivery and the `trap` registering). So, we can force the window to be bigger by putting a sleep before the `trap` registration. A `sleep 1` is probably big enough to always show the problem - but why not make the sleep huge (`sleep 10` for instance) to almost ensure we will see the issue all the time - just to prove a point :-)

Bottom line - this test was not showing a bug in the runtime - it is 'expected' behavior - if you hit the very small race window. The fix is to remove the race window from the test.

**Comment by sboeuf** [comment]
Thanks @grahamwhaley, all of this makes sense to me :)
